docs/rfc/rfc-2.adoc

```diff
@@ -0,0 +1,799 @@
+for closer monitoring by operators, ensuring these logs will be noticed. The
+descriptive prefix will further allow for targeted notifications by operators
+so that they can flag these logs for follow-up action.
+
+The suggested structure of the log message is:
+
+```
+2020-02-01T12:12:12.12-0800    WARN   keep-ecdsa   Please broadcast Bitcoin transaction 0010010101000828181124af234gbawed000
+```
+
+The first three components are the standard components emitted by the ECDSA
+client's logging library, while the message itself begins with `Please broadcast
+Bitcoin transaction`, followed by the hexadecimal string representing the signed
+transaction. The signed transaction can be broadcast via manual submission to a
+known node, via API access to a service provider like BlockCypher or
+Blockstream, or via a manual form submission on a page like
+https://live.blockcypher.com/btc/pushtx/ .
+
+If an optional Bitcoin node connection is available (see <<electrs>>), the
+client should attempt to publish the transaction by performing an HTTP POST
+request to path `/tx`. The POST request body should be the signed transaction
+hex, as would be logged in the WARN logs above. If the POST fails, the client
+should log an error and then fall back on the WARN logging above.
+
+=== Implementation Details
+
+There are a few implementation details that are worth calling out; in
+particular, the *pub address derivation, the deposit output script
+construction, the Bitcoin sighash construction, the Bitcoin transaction
+construction, and the attachment of the signature to the Bitcoin transaction.
+This proposal suggests this be done using the `btcsuite` series of Go
+libraries, and each one is detailed below with draft code using those
+libraries.
+
+[[output-script]]
+==== Deposit Output Script Construction
+
+tBTC deposits always use Witness PubKeyHash in their funding UTXO. This means
+that, to spend the funding UTXO, the deposit signing group must sign a sighash
+that allows spending that UTXO. The signing process does this by including a
+`scriptCode` as part of the the sighash (see <<sighash-construction>>). The
+`scriptCode`, for Witness PubKeyHash transactions, is defined in BIP143.
+
+This function constructs the relevant `scriptCode` bytes from the deposit's
+public key:
+
+```go
+// import "github.com/btcsuite/btcd/btcec"
+// import "github.com/btcsuite/btcutil"
+
+// PublicKeyToP2WPKHOutputScript converts a public key to a Bitcion p2wpkh
+// witness scriptCode that can spend an output sent to that public key's
+// corresponding address.
+//
+// [BIP141]: https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki
+func PublicKeyToP2WPKHScriptCode(
+  publicKey *ecdsa.PublicKey,
+  chainParams *chaincfg.Params,
+) ([]byte, error) {
+  // ecdsa.PublicKey and btcec.PublicKey are both method attachments to
+  // Go's crypto/ecdsa.PublicKey, so we can cast.
+	publicKeyBytes := (*btcec.PublicKey)(publicKey).SerializeCompressed()
```

Comment:
Been there, done that. 🔨 Still hurts.

Reply:
It doesn't mean I have any better idea, just expressing my support.

Reply:
Yeah it's monstrous.

Reply:
I did steal it from you though :grin:

Reply:
apes together strong